### PR TITLE
Part7to9

### PR DESCRIPTION
**Pull request by hizkiasusanto**
Part 7 to 9

### CODE REVIEW

File: Chessington.GameEngine/Pieces/King.cs [hunk 1]
```diff
@@ -6,11 +6,32 @@ namespace Chessington.GameEngine.Pieces
...
+                    if (Square.At(currentRow + rowDiff, currentCol + colDiff).IsInbound())
+                    {
+                        availableMoves.Add(Square.At(currentRow + rowDiff, currentCol + colDiff));
+                    }
+                }
+            }
+
+            availableMoves.RemoveAll(x => x == currentPos);
+
+            availableMoves.OrderBy(x => x.Row).OrderBy(x => x.Col);
```
Comment (ffarmanov) [inline Chessington.GameEngine/Pieces/King.cs:33]: Does it have to be ordered?

File: Chessington.GameEngine/Pieces/Knight.cs [hunk 2]
```diff
@@ -10,7 +10,30 @@ public Knight(Player player)
 
         public override IEnumerable<Square> GetAvailableMoves(Board board)
```
Comment (ffarmanov) [inline Chessington.GameEngine/Pieces/Knight.cs]: please look for 6.2 pull request for the comments

File: Chessington.GameEngine/Pieces/Pawn.cs [hunk 3]
```diff
@@ -17,14 +17,29 @@ public override IEnumerable<Square> GetAvailableMoves(Board board)
             
             if (Player == Player.White)
             {
-                availableMoves.Add(new Square(currentPosition.Row - 1, currentPosition.Col));
-                if (currentPosition.Row == 6) availableMoves.Add(new Square(currentPosition.Row -2, currentPosition.Col));
+                if (!Square.At(currentPosition.Row - 1, currentPosition.Col).IsOccupied(board))
```
Comment (ffarmanov) [inline Chessington.GameEngine/Pieces/Pawn.cs]: I guess playing with white or black pieces is just a matter of a direction. Maybe you can identify a direction and do smth like `currentPosition.Row - 1*direction` ?

File: Chessington.GameEngine/Pieces/Piece.cs [hunk 4]
```diff
@@ -34,27 +34,49 @@ public IEnumerable<Square> GetDiagonalMoves(Board board)
...
             var currentPos = board.FindPiece(this);
-            var row = currentPos.Row;
-            var col = currentPos.Col;
+            var currentRow = currentPos.Row;
+            var currentCol = currentPos.Col;
             List<Square> availableMoves = new List<Square>();
-            foreach (var i in Enumerable.Range(0, 8))
+
+            int tileDistance = 1;
+            while (Square.At(currentRow + tileDistance * rowDirection, currentCol + tileDistance * colDirection)
```
Comment (ffarmanov) [inline Chessington.GameEngine/Pieces/Piece.cs:70]: maybe you can combine Inbound and IsOccupied into one method IsValid?

File: Chessington.GameEngine/Pieces/Piece.cs [hunk 5]
```diff
@@ -34,27 +34,49 @@ public IEnumerable<Square> GetDiagonalMoves(Board board)
...
             availableMoves.RemoveAll(x => x == Square.At(currentRow, currentCol));
-            return availableMoves; 
+            return availableMoves;
         }
 
         public IEnumerable<Square> GetLateralMoves(Board board)
+        {
+            List<Square> availableMoves = new List<Square>();
+
+            availableMoves.AddRange(ExploreInOneDirection(1, 0, board));
```
Comment (ffarmanov) [inline Chessington.GameEngine/Pieces/Piece.cs:52]: nice one, I like it

File: Chessington.GameEngine/Square.cs [hunk 6]
```diff
@@ -49,5 +51,22 @@ public override string ToString()
         {
             return string.Format("Row {0}, Col {1}", Row, Col);
         }
+
+        public bool IsInbound()
+        {
+            return Enumerable.Range(0, 8).Contains(this.Col) && Enumerable.Range(0, 8).Contains(this.Row);
```
Comment (ffarmanov) [inline Chessington.GameEngine/Square.cs]: 8 and similar numbers in a middle of code are called "Magic numbers", You should assign in to a constant and reuse it. 
In our case it's already a constant (refer to GameSettings class)

File: Chessington.GameEngine/Square.cs [hunk 7]
```diff
@@ -49,5 +51,22 @@ public override string ToString()
...
+        }
+
+        public bool IsOccupied(Board board)
+        {
+            if (board.GetPiece(this) == null)
+            {
+                return false;
```
Comment (ffarmanov) [inline Chessington.GameEngine/Square.cs]: you don't need to return false or true. you can just return the statement in if clause

**Review by ffarmanov — CHANGES_REQUESTED** [review]
Nice one, good progress 👍

File: Chessington.GameEngine/Pieces/Pawn.cs (hunk 3 above)
Comment (hizkiaadrian) [reply Chessington.GameEngine/Pieces/Pawn.cs:20]: This has been changed in the next pull request (10-12)

File: Chessington.GameEngine/Pieces/Piece.cs (hunk 4 above)
Comment (hizkiasusanto) [reply Chessington.GameEngine/Pieces/Piece.cs:70]: This has been changed in the next pull request (10-12)

**Comment by ffarmanov** [comment]
looks good you can merge it, just change RemoveAll(!Inbound) to Where(Inbound)

**Review by ffarmanov — APPROVED** [review]

**Pull request merged**